### PR TITLE
IOperation Test Leg

### PR DESCRIPTION
**Pull request by 333fred**
Adds a test leg that verifies IOperation in CI. Fixes https://github.com/dotnet/roslyn/issues/31450.

### CODE REVIEW

File: eng/build.ps1 [hunk 1]
```diff
@@ -94,6 +94,7 @@ function Print-Usage() {
   Write-Host "  -test64                   Run units tests in the 64-bit runner"
   Write-Host "  -testDesktop              Run Desktop unit tests (short: -test)"
   Write-Host "  -testCoreClr              Run CoreClr unit tests"
+  Write-Host "  -testCompilerOnly         Run only the compiler unit tests"
```
Comment (RikkiGibson) [inline eng/build.ps1:97]: Good idea.

File: src/Compilers/Test/Core/Compilation/CompilationExtensions.cs [hunk 2]
```diff
@@ -30,7 +32,13 @@ namespace Microsoft.CodeAnalysis.Test.Utilities
 {
     public static class CompilationExtensions
     {
-        internal static bool EnableVerifyIOperation { get; } = !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("ROSLYN_TEST_IOPERATION"));
+        internal static bool EnableVerifyIOperation { get; } =
+#if ROSLYN_TEST_IOPERATION
```
Comment (RikkiGibson) [inline src/Compilers/Test/Core/Compilation/CompilationExtensions.cs:36]: what benefit does this offer? is it just found to be easier to introduce the `#define` for local development versus adding an environment variable to the local machine or in IDE?

Comment (RikkiGibson) [reply src/Compilers/Test/Core/Compilation/CompilationExtensions.cs:36]: also, I realize this is still in draft mode, so.. feel free to ignore this.

Comment (333fred) [reply src/Compilers/Test/Core/Compilation/CompilationExtensions.cs:36]: > is it just found to be easier to introduce the #define for local development versus adding an environment variable to the local machine or in IDE?

Absolutely. You have to restart VS to get it to see the environment variable.

File: src/Tools/Source/RunTests/TestRunner.cs [hunk 3]
```diff
@@ -158,15 +158,19 @@ string makeHelixWorkItemProject(AssemblyInfo assemblyInfo)
                 var rehydrateFilename = isUnix ? "rehydrate.sh" : "rehydrate.cmd";
                 var lsCommand = isUnix ? "ls" : "dir";
                 var rehydrateCommand = isUnix ? $"./{rehydrateFilename}" : $@"call .\{rehydrateFilename}";
-                var workItem = @"
-        <HelixWorkItem Include=""" + assemblyInfo.DisplayName + @""">
-            <PayloadDirectory>" + Path.Combine(msbuildTestPayloadRoot, Path.GetDirectoryName(assemblyInfo.AssemblyPath)!) + @"</PayloadDirectory>
+                var setEnvironmentVariables = Environment.GetEnvironmentVariable("ROSLYN_TEST_IOPERATION") is { } iop
+                    ? $"{(isUnix ? "export" : "set")} ROSLYN_TEST_IOPERATION = {iop}"
```
Comment (RikkiGibson) [inline src/Tools/Source/RunTests/TestRunner.cs:162]: nit: I think there can't be spaces around the `=` in bash.

```bash
$ export a = b
bash: export: `=': not a valid identifier
```

**Comment by 333fred** [comment]
@jaredpar @RikkiGibson please review.

File: azure-pipelines.yml [hunk 4]
```diff
@@ -103,6 +103,15 @@ jobs:
...
     testArguments: -testCoreClr
 
+- template: eng/pipelines/test-windows-job.yml
+  parameters:
+    testRunName: 'Test Windows CoreCLR IOperation Debug'
+    jobName: Test_Windows_CoreClr_IOperation_Debug
+    buildJobName: Build_Windows_Debug
+    testArtifactName: Transport_Artifacts_Windows_Debug
+    configuration: Debug
+    testArguments: -testCoreClr -testIOperation -testCompilerOnly
```
Comment (RikkiGibson) [inline azure-pipelines.yml:113]: did you observe a difference in the number of tests run between `-testCoreClr` and `-testCoreClr -testCompilerOnly`?

**Review by RikkiGibson — APPROVED** [review]

File: eng/build.ps1 [hunk 5]
```diff
@@ -366,14 +383,23 @@ function TestUsingRunTests() {
...
-    $args += " --include '\.UnitTests'"
     $args += " --timeout 90"
+    if ($testCompilerOnly) {
+      $args += GetCompilerTestAssembliesIncludePaths
+    } else {
+      $args += " --include '\.UnitTests'"
+    }
   }
-  elseif ($testDesktop -or $testIOperation) {
+  elseif ($testDesktop -or ($testIOperation -and -not $testCoreClr)) {
```
Comment (RikkiGibson) [inline eng/build.ps1:393]: this additional logic looks like it's meant to make it so you only need to pass the single `.\Test.cmd -testIOperation` flag to run ioperation tests locally. Is that right?

Comment (333fred) [reply eng/build.ps1:393]: Correct, that's the behavior we have today.

File: azure-pipelines.yml (hunk 4 above)
Comment (333fred) [reply azure-pipelines.yml:113]: Yes, looking at the pipeline it runs about 4k less tests.

**Comment by 333fred** [comment]
@jaredpar for a second review.

**Comment by 333fred** [comment]
@jaredpar please take look.

File: eng/build.ps1 [hunk 6]
```diff
@@ -327,6 +329,21 @@ function GetIbcDropName() {
...
+  $assemblies += " --include '^Microsoft\.CodeAnalysis\.CSharp\.Syntax\.UnitTests$'"
+  $assemblies += " --include '^Microsoft\.CodeAnalysis\.CSharp\.Symbol\.UnitTests$'"
+  $assemblies += " --include '^Microsoft\.CodeAnalysis\.CSharp\.Semantic\.UnitTests$'"
+  $assemblies += " --include '^Microsoft\.CodeAnalysis\.CSharp\.Emit\.UnitTests$'"
+  $assemblies += " --include '^Microsoft\.CodeAnalysis\.CSharp\.IOperation\.UnitTests$'"
+  $assemblies += " --include '^Microsoft\.CodeAnalysis\.VisualBasic\.Syntax\.UnitTests$'"
+  $assemblies += " --include '^Microsoft\.CodeAnalysis\.VisualBasic\.Symbol\.UnitTests$'"
+  $assemblies += " --include '^Microsoft\.CodeAnalysis\.VisualBasic\.Semantic\.UnitTests$'"
+  $assemblies += " --include '^Microsoft\.CodeAnalysis\.VisualBasic\.Emit\.UnitTests$'"
+  $assemblies += " --include '^Roslyn\.Compilers\.VisualBasic\.IOperation\.UnitTests$'"
```
Comment (jaredpar) [inline eng/build.ps1:343]: Shouldn't this be named MS.CA.VB.IOperation.UnitTests? Not something to follow up on in this PR but seems like a name we should be fixing at some point.

File: eng/build.ps1 [hunk 7]
```diff
@@ -327,6 +329,21 @@ function GetIbcDropName() {
     return $drop.Name
 }
 
+function GetCompilerTestAssembliesIncludePaths() {
```
Comment (jaredpar) [inline eng/build.ps1:332]: This is excluding the following compiler tests:
- Command Line
- Server

File: eng/build.ps1 [hunk 8]
```diff
@@ -366,14 +383,23 @@ function TestUsingRunTests() {
   if ($testCoreClr) {
     $args += " --tfm net5.0"
     $args += " --tfm netcoreapp3.1"
-    $args += " --include '\.UnitTests'"
     $args += " --timeout 90"
+    if ($testCompilerOnly) {
+      $args += GetCompilerTestAssembliesIncludePaths
+    } else {
+      $args += " --include '\.UnitTests'"
```
Comment (jaredpar) [inline eng/build.ps1:390]:
```suggestion
      $args += " --include '\.UnitTests$'"
```

Consistency with the other regex in this change.

File: eng/build.ps1 [hunk 9]
```diff
@@ -366,14 +383,23 @@ function TestUsingRunTests() {
...
-  elseif ($testDesktop -or $testIOperation) {
+  elseif ($testDesktop -or ($testIOperation -and -not $testCoreClr)) {
     $args += " --tfm net472"
-    $args += " --include '\.UnitTests'"
     $args += " --timeout 90"
 
+    if ($testCompilerOnly) {
+      $args += GetCompilerTestAssembliesIncludePaths
+    } else {
+      $args += " --include '\.UnitTests'"
```
Comment (jaredpar) [inline eng/build.ps1:400]:
```suggestion
      $args += " --include '\.UnitTests$'"
```

Consistency with other regex in the change.

File: eng/build.ps1 (hunk 8 above)
Comment (333fred) [reply eng/build.ps1:390]: I haven't gone through all the test projects in Roslyn to ensure that they end in `UnitTests`, and the existing regex used this format. I'd prefer to keep it as is.

File: eng/build.ps1 (hunk 6 above)
Comment (333fred) [reply eng/build.ps1:343]: Yes, you can blame my scatter-brainess from when I split out the unit tests a few years ago. I've always meant to go do a rename, just has never been on my priority list.

**Comment by 333fred** [comment]
@jaredpar addressed feedback

**Review by jaredpar — APPROVED** [review]

**Pull request merged**